#557 Add documentation about using fields as read/write accessors

filiphr opened the pull request:
This is a separate PR for the documentation for the new field mappings ability introduced in #928

agudian commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -350,14 +357,31 @@ public class CarMapperImpl implements CarMapper {
 
 MapStruct takes all public properties of the source and target types into account. This includes properties declared on super-types.
 
+[[field-mappings]]
```
Let's put this section to the end of chapter 3 (below chapter `[[updating-bean-instances]]`. It's a feature several people have asked for, but I guess many users don't use public field access and therefor we should treat it more as an advanced "hey by the way, we can also do this" thing.

filiphr commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -350,14 +357,31 @@ public class CarMapperImpl implements CarMapper {
 MapStruct takes all public properties of the source and target types into account. This includes properties declared on super-types.
 
+[[field-mappings]]
+=== Mappings via field
```
Much better like that 😄, I'll change it

filiphr commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -350,14 +357,31 @@ public class CarMapperImpl implements CarMapper {
+considered as a write accessor.
+
+You can find a complete example in the
+https://github.com/mapstruct/mapstruct-examples/tree/master/mapstruct-field-mapping[mapstruct-examples-field-mapping]
```
The examples I wrote have exactly that. There is one bean with public fields, the target bean with getters and setters (This would be the use case that I see people using the most). Will add those to the documentation as well.

filiphr commented:
Thanks for the review @agudian. I removed all the parts that you mentioned, changed the section name and added code snippets. Do you think that their length is too much? Should I reduce them?

agudian commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -500,6 +500,163 @@ The generated code of the `updateCarFromDto()` method will update the passed `Ca
+
+    private Long id;
+    private String name;
+    private Collection<OrderItem> orderItems;
```
I guess the order item isn't really necessary to demonstrate the feature here.

agudian commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -500,6 +500,163 @@ The generated code of the `updateCarFromDto()` method will update the passed `Ca
+
+    CustomerMapper MAPPER = Mappers.getMapper( CustomerMapper.class );
+
+    @Mapping(source = "orders", target = "orderItems")
```
Without the order, you could map `name` to `customerName` or something?

agudian commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -500,6 +500,163 @@ The generated code of the `updateCarFromDto()` method will update the passed `Ca
+        customer.setName( customerDto.name );
+
+        return customer;
+    }
```
You could abbreviate this to:
```java
// GENERATED CODE
public class CustomerMapperImpl implements CustomerMapper {

    @Override
    public Customer toCustomer(CustomerDto customerDto) {
        // ... 
        customer.setId( customerDto.id );
        customer.setName( customerDto.name );
        // ...
    }

    @Override
    public CustomerDto fromCustomer(Customer customer) {
        // ...
        customerDto.id = customer.getId();
        customerDto.name = customer.getName();
        // ...
    }
}
```

filiphr commented:
You are right. It is much better now. Thanks

agudian commented:
👍 go ahead and squash/merge!

filiphr commented:
@agudian Is the last addition with the "experimental" warning OK for you?

sjaakd commented:
@filiphr .. is there a reason to mark this as `experimental`? I think the beta upcoming beta release of 1.2 already expresses this.

Which triggers me.. perhaps we should have some basic guidelines what we regard as experimental. Up to now (IMHO) these are features for which we're not sure about their permanent character. Typically we discuss long on those. For myself, this is not a debatable one.

filiphr commented:
@sjaakd only because it is new. However, you are right the beta release expresses it and it will be in the release notes. I will remove that from there.

>Which triggers me.. perhaps we should have some basic guidelines what we regard as experimental. Up to now (IMHO) these are features for which we're not sure about their permanent character. Typically we discuss long on those. For myself, this is not a debatable one.

Setting up some basic guidelines would be great.

gunnarmorling commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -500,6 +500,86 @@ The generated code of the `updateCarFromDto()` method will update the passed `Ca
+use the fields as read/write accessor if it cannot find suitable getter/setter methods for the property.
+
+A field is considered as a read accessor if it is `public` or `public final`. If a field is `static` it is not
+considered as a read accessors.
```
accessors=accessor

filiphr replied:
We missed that one. Should I just push to master with this change, or create a new PR? This PR is already merged 😄

gunnarmorling commented on documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc:
```diff
@@ -500,6 +500,86 @@ The generated code of the `updateCarFromDto()` method will update the passed `Ca
+[[direct-field-mappings]]
+=== Mappings with direct field access
+
+MapStruct also supports mappings of `public` fields that have no getters/setters. MapStruct will
```
Does it support package visible accessors, too?

filiphr replied:
No it doesn't it only accepts `public` visible accessors. I think that we agreed adding the ability to use `package`/`protected` accessors as part of #572?

gunnarmorling commented:
Concering the "experimental" status, the idea is to use it in a final release (i.e. non-Beta) if the feature still may change as we need more feedback and real-world usage feedback.

gunnarmorling commented:
Just push an update to master. Thanks! Sorry for commenting on the closed
PR, I'm a bit late to the party :)

gunnarmorling commented:
>  I think that we agreed adding the ability to use package/protected accessors
as part of #572 <https://github.com/mapstruct/mapstruct/issues/572>?

Ah, got it. That's cool.

filiphr commented:
>Just push an update to master. Thanks! Sorry for commenting on the closed
PR, I'm a bit late to the party :)

No problem. It is not nice to have typos and other problems. Fixed it in f949902
